### PR TITLE
Add CPU optimized kernels for topk and rope fusions 

### PR DESCRIPTION
**Pull request by jianan-gu**
## Overview:

This PR is adding the following CPU optimized sgl-kernels that are at least used in Qwen3/LLama1-4 models:
```
- TopK fusions: 
      TopK+sigmoid
      softmax+TopK
- Norm fusion:
      L2norm
- RoPE fusions:
      origin rope fusion (gpt_neox style and gptj style)
```

### CODE REVIEW

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 1]
```diff
@@ -65,8 +65,230 @@ void rope_kernel_impl(
...
+  }
+}
+
+template <typename scalar_t>
+void rotary_embedding_origin_gptj_impl(
+    const int64_t* __restrict__ positions,       // [batch_size, seq_len] or
+                                                 // [num_tokens]
+    scalar_t* __restrict__ query,                /// [batch_size, seq_len, num_heads,
+                                                 /// head_size] or [num_tokens, num_heads,
+                                                 /// head_size]
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/rope.cpp]:
```suggestion
    scalar_t* __restrict__ query,                // [batch_size, seq_len, num_heads,
                                                 // head_size] or [num_tokens, num_heads,
                                                 // head_size]
```

File: test/srt/cpu/test_moe_topk_softmax.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import itertools
+
+import pytest
+import sgl_kernel
+import torch
+
+torch.manual_seed(1998)
```
Comment (mingfeima) [inline test/srt/cpu/test_moe_topk_softmax.py]: set the seed in function.

File: test/srt/cpu/test_moe_topk_softmax.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    assert torch.allclose(
+        topk_weights_ref.to(test_dtype),
+        topk_weights.to(test_dtype),
+    ), f"Weights mismatch: torch={topk_weights_ref} vs SGLang={topk_weights}"
+
+    assert torch.allclose(
+        topk_indices_ref.sort()[0].to(torch.int32)[0, :],
+        topk_indices.sort()[0].to(torch.int32)[0, :],
+        atol=0,
+        rtol=0,
```
Comment (mingfeima) [inline test/srt/cpu/test_moe_topk_softmax.py]: scatter back topk_weights with regard to topk_indices and compare, take test_topk.py as a reference.

we may have identical values in topk_weights and different topk_indices.

File: sgl-kernel/csrc/cpu/topk.cpp [hunk 4]
```diff
@@ -136,6 +136,17 @@ void grouped_topk_kernel_impl(
   });
 }
 
+inline void _sigmoid(float* __restrict__ out, const float* __restrict__ input, int SIZE) {
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/topk.cpp]: is this function used?

File: sgl-kernel/csrc/cpu/topk.cpp [hunk 5]
```diff
@@ -157,6 +168,103 @@ inline void sigmoid(float* __restrict__ out, const scalar_t* __restrict__ input)
   }
 }
 
+template <typename scalar_t, int SIZE>
+inline void _copy_and_convert_to_float(float* __restrict__ out, const scalar_t* __restrict__ input) {
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/topk.cpp]: you can use at::vec::convert<scalar_t, float>, it is vectorized.

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 6]
```diff
@@ -65,8 +65,230 @@ void rope_kernel_impl(
...
+    const int rot_dim,
+    const int64_t query_stride,
+    const int64_t key_stride,
+    const int num_heads,
+    const int num_kv_heads,
+    const int head_size,
+    const int num_tokens) {
+  const int embed_dim = rot_dim / 2;
+
+#pragma omp parallel for collapse(2)
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/rope.cpp]: use at::parallel_for.

i will replace the parallel wrapper later on. So keep it in uniform here.

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 7]
```diff
@@ -65,8 +65,230 @@ void rope_kernel_impl(
...
+    for (int i = 0; i < num_kv_heads; ++i) {
+      const int head_idx = i;
+      const int64_t token_head = token_idx * key_stride + head_idx * head_size;
+      compute_loop(token_head, cache_ptr, key);
+    }
+  }
+}
+
+template <typename scalar_t>
+void rotary_embedding_origin_gptj_impl(
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/rope.cpp]: can we merge this function with the existing one. say pass a lambda as an arguement.

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 8]
```diff
@@ -65,8 +65,230 @@ void rope_kernel_impl(
...
   });
 }
+
+template <typename scalar_t>
+void rotary_embedding_origin_impl(
+    const int64_t* __restrict__ positions,       // [batch_size, seq_len] or
+                                                 // [num_tokens]
+    scalar_t* __restrict__ query,                /// [batch_size, seq_len, num_heads,
+                                                 /// head_size] or [num_tokens, num_heads,
+                                                 /// head_size]
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/rope.cpp]:
```suggestion
    scalar_t* __restrict__ query,                // [batch_size, seq_len, num_heads,
                                                 // head_size] or [num_tokens, num_heads,
                                                 // head_size]
```

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 9]
```diff
@@ -65,8 +65,230 @@ void rope_kernel_impl(
     }
   });
 }
+
+template <typename scalar_t>
+void rotary_embedding_origin_impl(
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/rope.cpp]: can we merge this function with the existing one. say pass a lambda as an argument.

**Review by mingfeima — CHANGES_REQUESTED** [review]
* merge test_moe_topk_sigmoid.py and test_moe_topk_softmax.py with existing one: test_topk.py
* try to merge the rope impls with existing one, say pass a inline function as argument at the entry function or use ENUM and template specification. both methods are OK. this will greatly simplify the code structure.
* use at::parallel_for

File: sgl-kernel/csrc/cpu/rope.cpp (hunk 1 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/rope.cpp]: Have refined.

File: test/srt/cpu/test_moe_topk_softmax.py (hunk 2 above)
Comment (jianan-gu) [reply test/srt/cpu/test_moe_topk_softmax.py]: Sure, have moved them in test functions

File: test/srt/cpu/test_moe_topk_softmax.py (hunk 3 above)
Comment (jianan-gu) [reply test/srt/cpu/test_moe_topk_softmax.py]: Sure, have refined test cases according to test_topk.py.

File: sgl-kernel/csrc/cpu/topk.cpp (hunk 4 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/topk.cpp]: no, have removed.

File: sgl-kernel/csrc/cpu/topk.cpp (hunk 5 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/topk.cpp]: refined accordingly

File: sgl-kernel/csrc/cpu/rope.cpp (hunk 6 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/rope.cpp]: have replace with at::parallel_for.

File: sgl-kernel/csrc/cpu/rope.cpp (hunk 7 above)
Comment (jianan-gu) [reply sgl-kernel/csrc/cpu/rope.cpp]: Detailed check current Deepseek rope and origin rope (gptj/neox style), following key differences of usages:

- input 3D for Deepseek rope (num tokens, heads, head size), while origin rope is with input 2D (num tokens, heads * head size), not spilt from Q/K
- Deepseek rope takes q_pe/ k_pe, all embedding dim is aligned with input dim and rope dim, while origin rope could accept different dims

Thus, merging functions from their implementations may not be strightforward and hard to be common for all. Here trying to refine rope to merge the dispatch ops together and leave only one `rotary_embedding_cpu` as common, while the underneath implementations remained different. And total 4 kinds of implementations are needed ( input 2D or 3D, is_neox or not)

File: sgl-kernel/csrc/cpu/rope.cpp [hunk 10]
```diff
@@ -4,126 +4,343 @@
...
+            num_kv_heads,
+            head_size,
+            num_tokens);
+      }
+      query_out = query;
+      key_out = key;
+
+    } else {
+      TORCH_CHECK(
+          is_neox == false, " Query/Key with 3D [num_tokens, num_heads, head_size] does not suport neox rope yet");
```
Comment (zhyncs) [inline sgl-kernel/csrc/cpu/rope.cpp]:
```suggestion
          is_neox == false, " Query/Key with 3D [num_tokens, num_heads, head_size] does not support neox rope yet");
```

**Review by zhyncs — APPROVED** [review]
